### PR TITLE
Scroll to the right spot when clicking "Fix Errors"

### PR DESCRIPTION
**Pull request by Gonals**
### Fixed Issues

$ https://github.com/Expensify/App/issues/12942

### Tests
1. Go to Settings => Payments => Add a debit card
2. Without entering any data, click "Save"
3. Click on 'fix the errors' button
4. System should focus on the first input and scroll to it with no lines cut
5. Populate the first field and make the screen small enough so that not all fields fit in it.
6. Click 'fix errors again'
7. You should now get scrolled to the second field (and it should get focused too)

- [x] Verify that no errors appear in the JS console

### Offline tests
Not relevant

### QA Steps
Same as Tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The reviewer will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for expected offline behavior are in the `Offline steps` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [ ] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] I verified that if a function's arguments changed that all usages have also been updated correctly
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### Screenshots/Videos
<details>
<summary>Web</summary>

Screen Shot 2022-11-23 at 1 18 07 PM

</details>

<details>
<summary>Desktop</summary>

Screen Shot 2022-11-23 at 4 43 50 PM

</details>

<details>
<summary>iOS</summary>

Screen Shot 2022-11-23 at 3 47 49 PM

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

File: src/components/Form.js [hunk 1]
```diff
@@ -245,6 +246,10 @@ class Form extends React.Component {
                     style={[styles.w100, styles.flex1]}
                     contentContainerStyle={styles.flexGrow1}
                     keyboardShouldPersistTaps="handled"
+                    ref={el => this.form = el}
+                    onLayout={(event) => {
+                        this.inputPosition = event.nativeEvent.layout.y + 1;
```
[the remark this answers is not in the document]
Comment (Gonals) [reply src/components/Form.js:253]: Yep, adding a comment. 

Without the +1, we'll scroll right to the border of the component, potentially covering the upper border

**Comment by Gonals** [comment]
Checks passing! I'm having trouble testing on Android today, so if you could give it a try there, it'd be awesome!

File: src/components/Form.js [hunk 2]
```diff
@@ -259,6 +266,9 @@ class Form extends React.Component {
                                 const errors = !_.isEmpty(this.state.errors) ? this.state.errors : this.props.formState.errorFields;
                                 const focusKey = _.find(_.keys(this.inputRefs), key => _.keys(errors).includes(key));
                                 this.inputRefs[focusKey].focus();
+                                this.inputRefs[focusKey].measure((fx, fy, width, height, px, py) => {
+                                    this.form.scrollTo({y: py - this.viewPosition, animated: false});
```
[the remark this answers is not in the document]
Comment (Gonals) [reply src/components/Form.js:270]: Oh, nice catch! Looking into it

[the remark this answers is not in the document]
Comment (Gonals) [reply src/components/Form.js:269]: Ok, I changed things around a little bit and I think it is working more smoothly now!
Quick thing, I see you added a bunch of `Name` fields for testing. Make sure you give them different `inputID`, as those are involved when determining where to scroll and might be the reason of the weird behavior you saw

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
> Quick thing, I see you added a bunch of Name fields for testing. Make sure you give them different inputID, as those are involved when determining where to scroll and might be the reason of the weird behavior you saw

thanks @Gonals , I'll make sure to use unique ids.

prerequisites: 
1. increase the page height. This can be done by inserting a bunch of dummy input fields just after name. make sure to use unique ids. 
2. use iOS and make sure the virtual keyboard is open by pressing `⌘ + K`

bug: focused input is behind the keyboard

steps:
1. click on zip code
2. click on expensify password
3. click on `fix the errors`
4. scroll down and click on `fix the errors` again.

Expected: zip code input is above the keyboard
Actual: zip code input is behind the keyboard 

https://user-images.githubusercontent.com/29673073/203745723-e116d31b-9ad2-41d0-95b7-ce3fd4f84db7.mov

**Comment by Gonals** [comment]
Ok! The math was becoming a pain, so I completely changed the approach! Ready for another look!

**Review by rushatgabhane — APPROVED** [review]

**Review by rushatgabhane — CHANGES_REQUESTED** [review]
@Gonals noticed three bugs on Android

1. Sometimes the form can't be scrolled on clicking fix errors for **expiry date** input (`0:10` mark in video)
2. On clicking fix errors, the password input is hidden behind the keyboard (`0:17` mark).
3. Click on password field and the save button is always behind the keyboard (`0:39` mark)

https://user-images.githubusercontent.com/29673073/204360635-70f0e4fe-3963-43bc-8437-ce146bbd336b.mp4

**Comment by Gonals** [comment]
> @Gonals noticed three bugs on Android
> 
> 1. Sometimes the form can't be scrolled on clicking fix errors for **expiry date** input (`0:10` mark in video)
> 2. On clicking fix errors, the password input is hidden behind the keyboard (`0:17` mark).
> 3. Click on password field and the save button is always behind the keyboard (`0:39` mark)
> 
>  WhatsApp.Video.2022-11-29.at.00.34.35.mp4

Hi @rushatgabhane,

1. This does not seem related to this PR. I haven't been able to reproduce it, but it may be device-specific. In any case, I don't think it is caused by this.
2.  I'm not sure I'm seeing this one in the video. The date is wrong, so that's where you are being scrolled to, right? That's the reason `password` gets covered
3. This one I can reproduce, but it happens in `main` too. Basically, any time `password` is selected. I don't think this is a bug, but, in any case, it is not related to this PR.

**Review by rushatgabhane — APPROVED** [review]
## Reviewer Checklist

@danieldoglas LGTM! 

@Gonals fantastic work in changing to a new approach! ✨


- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.


### Screenshots/Videos
<details>
<summary>Web</summary>


https://user-images.githubusercontent.com/29673073/204901525-803ec9c9-0275-46cb-a820-dec992bd03b9.mov



</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/29673073/204909392-479e2a26-bae4-4185-8957-92bfb8655a1a.mov



</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/29673073/204908717-99a2ade4-399d-4fcd-bb7a-35c0d9508e8c.mov



</details>

<details>
<summary>Desktop</summary>


https://user-images.githubusercontent.com/29673073/204901635-0b127018-fa2e-48e2-bee1-50056f6a39bd.mov



</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/29673073/204909663-e3a355c9-bdef-4e92-ad4c-4384cd80f219.mov



</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/29673073/204909454-98c5f03c-7297-4c44-bebc-e5a170dee9cf.mp4



</details>

**Comment by Gonals** [comment]
@danieldoglas All yours!

**Comment by Gonals** [comment]
@danieldoglas bump!

**Comment by danieldoglas** [comment]
oops, sorry! Looking into this now!

**Review by danieldoglas — APPROVED** [review]
code LGTM

**Comment by luacmartins** [comment]
@Gonals I'm seeing this error on the VBA page. See this issue https://github.com/Expensify/App/issues/13344
Screenshot_1670283345

The text elements don't have any props.

**Comment by aldo-expensify** [comment]
I think this PR caused this reggresison (not yet in staging): https://github.com/Expensify/App/issues/13401

If I revert the changes, the spinner is back at the middle of the page.

File: src/components/Form.js [hunk 3]
```diff
@@ -108,6 +109,16 @@ class Form extends React.Component {
...
     }
 
+    setPosition(element, position) {
+        if (!element.props.inputID && element.props.children) {
+            _.forEach(element.props.children, (child) => {
+                this.setPosition(child, position);
+            });
+        } else {
+            this.childPosition[element.props.inputID] = position;
+        }
```
Comment (aldo-expensify) [inline src/components/Form.js:119]: If `element.props.inputID` is false, but `element.props.children` is also falsey, then you will enter  in the `else` which I think is wrong.

Comment (aldo-expensify) [reply src/components/Form.js:119]: I put some console logs, and it doesn't happen in the debit card form, but in other forms like opening Personal details it does happen:

`element.props.inputID` is `undefined` and `element.props.children` is also `undefined`

**Comment by aldo-expensify** [comment]
Reverting PR here + different fix to achieve the same: https://github.com/Expensify/App/pull/13499

**Comment by luacmartins** [comment]
Just FYI that this PR caused [this bug](https://github.com/Expensify/App/issues/13344).

**Pull request merged**